### ISSUE TITLE
blt@0.3.6 unsatisfiable conflict w/ +rocm: umpire, raja, more?

### ISSUE THREAD

**Issue by eugeneswalker** [LLNL/spack#23473]
Using original concretizer from `spack@develop` (dfcd01f2c7ebbd0c1ac214c5c73ebc00b46d930d from `Thu May 6 12:11:20 2021 +0200`)

```
$> spack spec raja+rocm~openmp
Input spec
--------------------------------
raja~openmp+rocm amdgpu_target=gfx906

Concretized
--------------------------------
==> Error: Conflicts in concretized spec "raja@0.13.0%gcc@9.3.0~cuda+examples+exercises~ipo~openmp+rocm+shared~tests amdgpu_target=gfx906 build_type=RelWithDebInfo cuda_arch=none arch=linux-ubuntu20.04-cascadelake/ujiyfzv"

List of matching conflicts for spec:

    raja@0.13.0%gcc@9.3.0~cuda+examples+exercises~ipo~openmp+rocm+shared~tests amdgpu_target=gfx906 build_type=RelWithDebInfo cuda_arch=none arch=linux-ubuntu20.04-cascadelake
        ^blt@0.3.6%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
            ^cmake@3.20.2%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-cascadelake
                ^ncurses@6.2%gcc@9.3.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-cascadelake
                    ^pkgconf@1.7.4%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                ^openssl@1.1.1k%gcc@9.3.0~docs+systemcerts arch=linux-ubuntu20.04-cascadelake
                    ^perl@5.32.1%gcc@9.3.0+cpanm+shared+threads arch=linux-ubuntu20.04-cascadelake
                        ^berkeley-db@18.1.40%gcc@9.3.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu20.04-cascadelake
                        ^gdbm@1.19%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^readline@8.1%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                    ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-ubuntu20.04-cascadelake
        ^camp@0.1.0%gcc@9.3.0~cuda~ipo+rocm~tests amdgpu_target=gfx906 build_type=RelWithDebInfo cuda_arch=none arch=linux-ubuntu20.04-cascadelake
            ^hip@4.1.0%gcc@9.3.0~ipo build_type=RelWithDebInfo patches=2a4190477b7d9206b9cd8d70770ba0bc007273cbe54772efb12f9ca2e37c0392,5581cfa93974e82077bbc0b3b6590d312b9f8bca4162e691b0a63afa25f0de4c,9dc7efbdb3d7e375d2cc410702e2b9686cc73be40aa8d494b26be8a4538f18cd arch=linux-ubuntu20.04-cascadelake
                ^comgr@4.1.0%gcc@9.3.0~ipo build_type=Release patches=4dda4b3dcd61df74050076881e102c6f4ec637ea8f35a3a4e7079b8a4fa9af4c arch=linux-ubuntu20.04-cascadelake
                    ^llvm-amdgpu@4.1.0%gcc@9.3.0~ipo+openmp build_type=Release arch=linux-ubuntu20.04-cascadelake
                        ^hwloc@2.4.1%gcc@9.3.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-ubuntu20.04-cascadelake
                            ^libpciaccess@0.16%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^libtool@2.4.6%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                    ^m4@1.4.18%gcc@9.3.0+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-ubuntu20.04-cascadelake
                                        ^libsigsegv@2.13%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^util-macros@1.19.1%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^libxml2@2.9.10%gcc@9.3.0~python arch=linux-ubuntu20.04-cascadelake
                                ^libiconv@1.16%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^xz@5.2.5%gcc@9.3.0~pic arch=linux-ubuntu20.04-cascadelake
                        ^libelf@0.8.13%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^perl-data-dumper@2.173%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^python@3.8.10%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-ubuntu20.04-cascadelake
                            ^bzip2@1.0.8%gcc@9.3.0~debug~pic+shared arch=linux-ubuntu20.04-cascadelake
                                ^diffutils@3.7%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^expat@2.3.0%gcc@9.3.0+libbsd arch=linux-ubuntu20.04-cascadelake
                                ^libbsd@0.11.3%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                    ^libmd@1.0.3%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^gettext@0.21%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-cascadelake
                                ^tar@1.34%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^libffi@3.3%gcc@9.3.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu20.04-cascadelake
                            ^sqlite@3.35.4%gcc@9.3.0+column_metadata+fts~functions~rtree arch=linux-ubuntu20.04-cascadelake
                            ^util-linux-uuid@2.36.2%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^z3@4.8.9%gcc@9.3.0+python arch=linux-ubuntu20.04-cascadelake
                            ^py-setuptools@50.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                    ^rocm-cmake@4.1.0%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-cascadelake
                    ^rocm-device-libs@4.1.0%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-cascadelake
                        ^texinfo@6.5%gcc@9.3.0 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-ubuntu20.04-cascadelake
                ^hip-rocclr@4.1.0%gcc@9.3.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu20.04-cascadelake
                    ^hsa-rocr-dev@4.1.0%gcc@9.3.0~ipo build_type=Release patches=71e6851d9be8113bfb8d71b66a3171e0d0401bae5e6f161c9e7fe32558261a46 arch=linux-ubuntu20.04-cascadelake
                        ^hsakmt-roct@4.1.0%gcc@9.3.0~ipo build_type=Release arch=linux-ubuntu20.04-cascadelake
                            ^numactl@2.0.14%gcc@9.3.0 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94,62fc8a8bf7665a60e8f4c93ebbd535647cebf74198f7afafec4c085a8825c006 arch=linux-ubuntu20.04-cascadelake
                                ^autoconf@2.69%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^automake@1.16.3%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                    ^mesa18@18.3.6%gcc@9.3.0+glx~llvm~opengles+osmesa patches=55a5611ca9fcbe8324c4d68a07b338134954ff12c5b122dc78ff376f012a1414 swr=none arch=linux-ubuntu20.04-cascadelake
                        ^binutils@2.36.1%gcc@9.3.0~gas~gold~headers~interwork~ld~libiberty~lto+nls+plugins arch=linux-ubuntu20.04-cascadelake
                        ^bison@3.7.4%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^flex@2.6.4%gcc@9.3.0+lex patches=09c22e5c6fef327d3e48eb23f0d610dcd3a35ab9207f12e0f875701c677978d3 arch=linux-ubuntu20.04-cascadelake
                            ^findutils@4.8.0%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^help2man@1.47.16%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^glproto@1.4.17%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^libx11@1.7.0%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^inputproto@2.3.2%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^kbproto@1.0.7%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^libxcb@1.14%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^libpthread-stubs@0.4%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^libxau@1.0.8%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                    ^xproto@7.0.31%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^libxdmcp@1.1.2%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                                ^xcb-proto@1.14.1%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^xextproto@7.3.0%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^xtrans@1.3.5%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^libxext@1.3.3%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                        ^py-mako@1.0.4%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                            ^py-markupsafe@1.1.1%gcc@9.3.0 arch=linux-ubuntu20.04-cascadelake
                ^rocminfo@4.1.0%gcc@9.3.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu20.04-cascadelake

1. "^blt@:0.3.6" conflicts with "raja+rocm"
```

Same with `umpire+rocm`:
```
$> spack spec umpire+rocm amdgpu_target=gfx906
Input spec
--------------------------------
umpire+rocm
...
1. "^blt@:0.3.6" conflicts with "umpire+rocm"
```

Where is this conflict coming from?

@white238 @haampie ?

**Comment by haampie** [comment]
there's no registered version of raja that supports blt@0.4:, probably the others need 0.4?

**Comment by eugeneswalker** [comment]
> there's no registered version of raja that supports blt@0.4:, probably the others need 0.4?

I don't know the technical reason
(A) why blt 0.3.6 can't be used with +rocm
(B) where the conflict is triggered in Spack

I don't see the conflict statements defined in the raja and umpire packages, so where is this coming from?

**Comment by white238** [comment]
It is here:  https://github.com/spack/spack/blob/2ba493cd630dfc44e69c32b6e31993fa949fc10b/lib/spack/spack/build_systems/rocm.py#L110

There have been  a lot of fixes in BLT related to rocm since 0.3.6 but I am not sure if any of them were hard requirements.  I don't have time right now to test if 0.3.6 works right now.  Can one of you try removing that and see if it works?  It might have been put there in error.  It is possible only one machine didn't work with 0.3.6.

**Comment by eugeneswalker** [comment]
> https://github.com/spack/spack/blob/2ba493cd630dfc44e69c32b6e31993fa949fc10b/lib/spack/spack/build_systems/rocm.py#L110
> 
> Can one of you try removing that and see if it works? It might have been put there in error.

Just tried. Errors:

```
  >> 112    clang-12: error: no such file or directory: ''-fPIC''
  >> 113    CMake Error at RAJA_generated_LockFreeIndexSetBuilders.cpp.o.cmake:143 (message):
     114      Error generating
     115      /tmp/root/spack-stage/spack-stage-raja-0.13.0-dd5m7fny65bipffjuydxan3fyw6youvj/spack-build-dd5m7fn/CMakeFiles/RAJA.dir/src/./R
            AJA_generated_LockFreeIndexSetBuilders.cpp.o
     116
     117
  >> 118    make[2]: *** [CMakeFiles/RAJA.dir/build.make:94: CMakeFiles/RAJA.dir/src/RAJA_generated_LockFreeIndexSetBuilders.cpp.o] Error 1
     119    make[2]: *** Waiting for unfinished jobs....
  >> 120    clang-12: error: no such file or directory: ''-fPIC''
  >> 121    clang-12: error: no such file or directory: ''-fPIC''
  >> 122    CMake Error at RAJA_generated_MemUtils_HIP.cpp.o.cmake:143 (message):
     123      Error generating
     124      /tmp/root/spack-stage/spack-stage-raja-0.13.0-dd5m7fny65bipffjuydxan3fyw6youvj/spack-build-dd5m7fn/CMakeFiles/RAJA.dir/src/./R
            AJA_generated_MemUtils_HIP.cpp.o
     125
     126
  >> 127    CMake Error at RAJA_generated_PluginStrategy.cpp.o.cmake:143 (message):
     128      Error generating
     129      /tmp/root/spack-stage/spack-stage-raja-0.13.0-dd5m7fny65bipffjuydxan3fyw6youvj/spack-build-dd5m7fn/CMakeFiles/RAJA.dir/src/./R
            AJA_generated_PluginStrategy.cpp.o
     130
     131
  >> 132    make[2]: *** [CMakeFiles/RAJA.dir/build.make:108: CMakeFiles/RAJA.dir/src/RAJA_generated_MemUtils_HIP.cpp.o] Error 1
  >> 133    make[2]: *** [CMakeFiles/RAJA.dir/build.make:115: CMakeFiles/RAJA.dir/src/RAJA_generated_PluginStrategy.cpp.o] Error 1
  >> 134    clang-12: error: no such file or directory: ''-fPIC''
  >> 135    CMake Error at RAJA_generated_DepGraphNode.cpp.o.cmake:143 (message):
     136      Error generating
     137      /tmp/root/spack-stage/spack-stage-raja-0.13.0-dd5m7fny65bipffjuydxan3fyw6youvj/spack-build-dd5m7fn/CMakeFiles/RAJA.dir/src/./R
            AJA_generated_DepGraphNode.cpp.o
```
